File: 2016/python/05.py
# Advent of Code 2016, day 5
# (c) aichingert

import hashlib

def get_hex(s,index):
    return hashlib.md5((s + str(index)).encode('utf-8')).hexdigest()

def part1(inp):
    n = 0
    ans = ""
    while True:
        n += 1
        s = get_hex(inp,n)
        con = False
        if s[:5] != "00000":
            continue
        ans += s[5]
        if len(ans) == 8:
            return ans
def part2(inp):
    n = 0
    ans = [""]*8
    while True:
        n += 1
        s = get_hex(inp,n)

        if s[:5] != "00000" or not s[5].isdigit():
            continue

        loc = int(s[5])

        if loc > 7 or loc < 0 or ans[loc] != "":
            continue
        ans[loc] = s[6]

        if "" not in ans:
            return "".join(ans)


inp = open("../input/05").read()
print("Part 1:", part1(inp))
print("Part 2:", part2(inp))

File: 2016/rust/05.rs
// Advent of Code 2016, day 5
// (c) aichingert

#[path="../../utils/rust/md5.rs"] mod md5;

fn part1(inp: &str) -> String {
    let mut ans = String::new();
    let mut n = 0;

    'outer: loop {
        n += 1;
        let s = md5::md5_utf8((inp.to_owned() + &n.to_string()).as_str()).chars().collect::<Vec<char>>();
        
        for ch in s[..5].iter() {
            if ch != &'0' { continue 'outer; }
        }

        ans.push(s[5]);

        if ans.len() > 7 {
            return ans;
        }
    }
}

fn part2(inp: &str) -> String {
    let mut ans: Vec<char> = vec!['_';8];
    let mut n = 0;

    'outer: loop {
        n += 1;
        let s = md5::md5_utf8((inp.to_owned() + &n.to_string()).as_str()).chars().collect::<Vec<char>>();
        
        for ch in s[..5].iter() {
            if ch != &'0' { continue 'outer; }
        }

        let n: usize = ((s[5] as u8) - 48) as usize;

        if n < 8 && ans[n] == '_' {
            ans[n] = s[6];
        }

        if !ans.contains(&'_') {
            return ans.into_iter().collect::<String>();
        }
    }
}

fn main() {
    let inp = std::fs::read_to_string("../input/05").unwrap();

    println!("Part 1: {}", part1(&inp));
    println!("Part 2: {}", part2(&inp));
}
